File: algorithm/leetcodeCpp/flatten-binary-tree-to-linked-list.cpp
114. Flatten Binary Tree to Linked List
Difficulty: Medium

Given a binary tree, flatten it to a linked list in-place.

For example,
Given

         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
click to show hints.

Hints:
If you notice carefully in the flattened tree, each node right child points to the next node of a pre-order traversal.

// Time:  O(n)
// Space: O(h)
// recursion
class Solution {
public:
    void flatten(TreeNode *root) {
        if (!root) {
          return;
        }
        if (root->left) {
          flatten(root->left);
        }
        if (root->right) {
            flatten(root->right);
        }
        TreeNode *right = root->right;
        root->right = root->left;
        root->left = nullptr;
        while (root->right) {
            root = root->right;
        }
        root->right = right;
    }
};

// Time:  O(n)
// Space: O(h)
// stack
class Solution {
public:
    void flatten(TreeNode* root) {
        if (!root) {
            return;
        }
        stack<TreeNode*> s;
        s.push(root);
        while (!s.empty()) {
            TreeNode *curr = s.top(); s.pop();
            if (curr->left) {
                TreeNode *left = curr->left;
                while (left->right) {
                    left = left->right;
                }
                left->right = curr->right;
                curr->right = curr->left;
                curr->left = NULL;
            }
            if (curr->right) {
                s.push(curr->right);
            }
        }
    }
};

File: data-structure-and-algorithm/java/flatten-binary-tree-to-linked-list.java
/**
 * Flatten Binary Tree to Linked List
 * 
 * Given a binary tree, flatten it to a linked list in-place.
 * 
 * For example,
 * Given
 * 
 *          1
 *         / \
 *        2   5
 *       / \   \
 *      3   4   6
 *  
 * 
 * The flattened tree should look like:
 * 
 *    1
 *     \
 *      2
 *       \
 *        3
 *         \
 *          4
 *           \
 *            5
 *             \
 *              6
 * click to show hints.
 * 
 * Hints:
 * If you notice carefully in the flattened tree, each node's right child points to the next node of a pre-order traversal
 */

import java.util.Stack;

public class Solution {
    public void flatten(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();

        if (root == null) {
            stack.push(root);
        }

        while (!stack.isEmpty()) {
            TreeNode curr = stack.pop();
            if (curr.left != null) {
                TreeNode leftTree = curr.left;
                while (leftTree.right) {
                    leftTree = leftTree.right;
                }
                leftTree.right = curr.right;
                curr.right = curr.left;
                curr.left = null;
            }
            if (curr.right != null) {
                stack.push(curr.right);
            }
        }
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}